### PR TITLE
[RFC] Don't use reflection and get a look and feel similar to flex

### PR DESCRIPTION
**Pull request by mantiz**
These changes eliminate the use of reflection in order to access the `configurator` of flex. Instead, it instantiates an own instance the same way flex does. Additionally, all recipes are collected during package install/uninstall operations and are applied after all packages have been installed/uninstalled *before* the "normal" recipes are applied. This gives us a look and feel similar to flex.

I also removed the `ExtraFlexDebug` and `Proxy` classes. I couldn't find any reason for the `ExtraFlexDebug` class and since reflection is no longer required, the `Proxy` class was no longer used anywhere. This also means that there are no longer any extra-flex events fired. I don't know if these were added for debugging purposes or if they are really required for anything.

Last but not least, I had to add `symfony/flex` to the `require` section again (was changed to `require-dev`). Without this change some used classes could not be found.

Like the title says, I created this PR as an RFC. In #1 you said the plugin is not (yet) stable because it uses reflection and since I now had a deeper insight of how the plugin works, I thought it should also be possible without reflection. So, I played around and this is the result.

If you don't like these changes, this is completely fine for me. I just thought since I made it anyway, I would propose the changes to you and you can decide if you like it or not. :)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -34,7 +34,7 @@
         "branch-alias": {
             "dev-master": "1.0-dev"
         },
-        "class": "Covex\\Composer\\ExtraFlexDebug",
```
Comment (covex-nn) [inline composer.json:37]: Composer creates plugin class via `eval`, see https://github.com/composer/composer/blob/master/src/Composer/Plugin/PluginManager.php. And plugin code become non-accessable to debug step by step with xdebug and PhpStorm

File: src/Composer/Proxy.php [hunk 2]
```diff
@@ -1,175 +0,0 @@
...
-            $eventName = "flex-" . strtolower($this->event) . "-" . $name;
-        } else {
-            $eventName = null;
-        }
-
-        $method = $this->getMethod($name);
-        if ($method) {
-            if (!is_null($eventName)) {
-                $preEvent = new Event("pre-" . $eventName, [ "arguments" => $arguments ]);
-                $this->eventDispacher->dispatch($preEvent->getName(), $preEvent);
```
Comment (covex-nn) [inline src/Composer/Proxy.php:97]: It was an experiment. There were an issue in `symfony/flex` about extending Flex. This fireing `pre-flex-configurator-install`, `post-flex-configurator-install` and `pre-flex-configurator-unconfigure` events, for example. Plugin does not use it

`$flexPlugin->configurator` and `$flexPlugin->dowloader` are replaced by instanceof Proxy somewhere in `ExtraFlexPlugin::decorate`, and after that Flex plugin operates with Proxy.

**Review by covex-nn — CHANGES_REQUESTED** [review]
When i started to work on this plugin, i hoped, that in the future Flex will become more extendable, and i tried not to copy code from Flex just not to support it. Also, i tried to create a plugin `for Flex`, not just for Composer. So, if we will decouple from Flex and if we will use our own `Options` for our own `Configurator` object, then it would break the main concept =)

But concept was wrong, i guess. Instead of extending Flex, it would be easier to fork `symfony/flex`, to create `mantiz/skeleton` or `covex-nn/skeleton`, use and support them, and update fork after flex updates.

File: composer.json (hunk 1 above)
Comment (mantiz) [reply composer.json:37]: Ah, ok, makes sense. Thanks.

**Comment by mantiz** [comment]
Thanks for your explanation. Helps me a lot to better understand the intention of the plugin. I remember reading an issue from flex where this was discussed and iirc there were some reasons for not making flex extensible but I can not remember at the moment. I will read it again. :)

A good enough working solution exists for me, so I will close this pull request. :)

**Pull request closed without merging**